match2/commons/match_group_display_bracket.lua

```diff
@@ -1,738 +1,745 @@
 end
 
-function _pxFix(height)
-    return height + (height == math.floor(height) and 0 or 0.2)
-end
+-- Connector lines between a match and its lower matches
+Bracket.propTypes.NodeLowerConnectors = Bracket.propTypes.NodeBody
+function Bracket.NodeLowerConnectors(props)
+    DisplayUtil.assertPropTypes(props, Bracket.propTypes.NodeLowerConnectors)
+    local match = props.matchesById[props.matchId]
+    local layout = props.layoutsByMatchId[props.matchId]
+    local config = props.config
+    local lowerMatches = match.bracketData.lowerMatches
+
+    local lowerLayouts = Array.map(
+        lowerMatches,
+        function(x) return props.layoutsByMatchId[x.matchId] end
+    )
+
+    -- Compute partial sums of heights of lower round matches
+    local heightSums = LuaUtils.math.partialSums(
+        Array.map(lowerLayouts, function(layout) return layout.height end)
+    )
+
+    -- Compute joints of connectors
+    local jointIxs = {}
+    local jointIxAbove = 0
+    for ix = math.ceil(#lowerMatches / 2), 1, -1 do
+        jointIxAbove = jointIxAbove + 1
+        jointIxs[lowerMatches[ix].opponentIx] = jointIxAbove
+    end
+    local jointIxBelow = 0
+    -- middle lower match is repeated if odd
+    for ix = math.floor(#lowerMatches / 2) + 1, #lowerMatches, 1 do
+        jointIxBelow = jointIxBelow + 1
+        jointIxs[lowerMatches[ix].opponentIx] = jointIxBelow
+    end
+    local jointCount = math.max(jointIxAbove, jointIxBelow)
+
+    --
+    local lowerConnectorsNode = mw.html.create('div')
+        :addClass('brkts-round-lower-connectors')
+        :css('--skip-round', match.bracketData.skipRound)
+
+    -- Draw connectors between lower round matches and this match
+    for ix, x in ipairs(lowerMatches) do
+        local lowerLayout = lowerLayouts[ix]
+        local leftTop = layout.lowerMarginTop + heightSums[ix] + lowerLayout.mid
+        local rightTop = layout.matchMarginTop + ((x.opponentIx - 1) + 0.5) * config.opponentHeight
+        local jointLeft = (config.roundHorizontalMargin - 2) * jointIxs[x.opponentIx] / (jointCount + 1)
+
+        local segment1Node = html.create('div'):addClass('brkts-line')
+            :css('height', config.lineWidth .. 'px')
+            :css('width', jointLeft + config.lineWidth / 2 .. 'px')
+            :css('left', '0')
+            :css('top', leftTop - config.lineWidth / 2 .. 'px')
+
+        local segment2Node = html.create('div'):addClass('brkts-line')
+            :css('height', math.abs(leftTop - rightTop) .. 'px')
+            :css('width', config.lineWidth .. 'px')
+            :css('top', math.min(leftTop, rightTop) .. 'px')
+            :css('left', jointLeft - config.lineWidth / 2 .. 'px')
+
+        local segment3Node = html.create('div'):addClass('brkts-line')
+            :css('height', config.lineWidth .. 'px')
+            :css('left', jointLeft - config.lineWidth / 2 .. 'px')
+            :css('right', '0')
+            :css('top', rightTop - config.lineWidth / 2 .. 'px')
```

Comment:
Duplicate code, use function?

Reply:
This is drawing line segments so it looks repetitive. I don't think there's a simple way to split this out into a function because I'm using different ways of specifying endpoints.